src/server/tablet_snap.rs

```diff
@@ -0,0 +1,526 @@
+                .worker_threads(DEFAULT_POOL_SIZE)
+                .after_start_wrapper(tikv_alloc::add_thread_memory_accessor)
+                .before_stop_wrapper(tikv_alloc::remove_thread_memory_accessor)
+                .build()
+                .unwrap(),
+            raft_router: r,
+            security_mgr,
+            cfg_tracker,
+            cfg: config,
+            sending_count: Arc::new(AtomicUsize::new(0)),
+            recving_count: Arc::new(AtomicUsize::new(0)),
+            engine: PhantomData,
+            limiter,
+        };
+        snap_worker
+    }
+
+    fn refresh_cfg(&mut self) {
+        if let Some(incoming) = self.cfg_tracker.any_new() {
+            let limit = if incoming.snap_max_write_bytes_per_sec.0 > 0 {
+                incoming.snap_max_write_bytes_per_sec.0 as f64
+            } else {
+                f64::INFINITY
+            };
+            self.limiter.set_speed_limit(limit);
+            info!("refresh snapshot manager config";
+            "speed_limit"=> limit);
+            self.cfg = incoming.clone();
+        }
+    }
+}
+
+pub struct SendStat {
+    key: TabletSnapKey,
+    total_size: u64,
+    elapsed: Duration,
+}
+
+impl<E, R> Runnable for TabletRunner<E, R>
+where
+    E: KvEngine,
+    R: RaftStoreRouter<E> + 'static,
+{
+    type Task = Task;
+
+    fn run(&mut self, task: Task) {
+        match task {
+            Task::Recv { stream, sink } => {
+                let task_num = self.recving_count.load(Ordering::SeqCst);
+                if task_num >= self.cfg.concurrent_recv_snap_limit {
+                    warn!("too many recving snapshot tasks, ignore");
+                    let status = RpcStatus::with_message(
+                        RpcStatusCode::RESOURCE_EXHAUSTED,
+                        format!(
+                            "the number of received snapshot tasks {} exceeded the limitation {}",
+                            task_num, self.cfg.concurrent_recv_snap_limit
+                        ),
+                    );
+                    self.pool.spawn(sink.fail(status));
+                    return;
+                }
+                SNAP_TASK_COUNTER_STATIC.recv.inc();
+
+                let snap_mgr = self.snap_mgr.clone();
+                let raft_router = self.raft_router.clone();
+                let recving_count = Arc::clone(&self.recving_count);
+                recving_count.fetch_add(1, Ordering::SeqCst);
+                let limit = self.limiter.clone();
+                let task = async move {
+                    let result = recv_snap(stream, sink, snap_mgr, raft_router, limit).await;
+                    recving_count.fetch_sub(1, Ordering::SeqCst);
+                    if let Err(e) = result {
+                        error!("failed to recv snapshot"; "err" => %e);
+                    }
+                };
+                self.pool.spawn(task);
+            }
+            Task::Send { addr, msg, cb } => {
```

Comment:
Where's the code to call Task::Send

Reply:
It will be called by `RaftClient` if the sending message type is the snapshot type , it will call `send_snapshot_sock`.